### DISCUSSION TITLE
3.0.0 build fails on Ubuntu 20.04 with libssl-dev installed

### DISCUSSION THREAD

**Discussion by gfyoung** [rbenv/ruby-build#1858] [Build failures]
Similar story as https://github.com/rbenv/ruby-build/issues/1215

Using `ruby-build` at https://github.com/rbenv/ruby-build/commit/02ce13c1a19358d8237d632ce88a180e253064bb

**OS (lsb_release -a)**

~~~
LSB Version:    core-11.1.0ubuntu2-noarch:security-11.1.0ubuntu2-noarch
Distributor ID: Ubuntu
Description:    Ubuntu 20.04.1 LTS
Release:        20.04
Codename:       focal
~~~

**libssl-dev (apt-cache policy libssl-dev)**

~~~
libssl-dev:
  Installed: 1.1.1f-1ubuntu2.1
  Candidate: 1.1.1f-1ubuntu2.1
  Version table:
 *** 1.1.1f-1ubuntu2.1 500
        500 http://archive.ubuntu.com/ubuntu focal-updates/main amd64 Packages
        500 http://security.ubuntu.com/ubuntu focal-security/main amd64 Packages
        100 /var/lib/dpkg/status
     1.1.1f-1ubuntu2 500
        500 http://archive.ubuntu.com/ubuntu focal/main amd64 Packages
~~~

**OpenSSL**

~~~
$ which openssl -a
/usr/bin/openssl

$ openssl version
OpenSSL 1.1.1f  31 Mar 2020
~~~

**Input**

~~~
ruby-build 3.0.0 ~/ruby-test
~~~

**Output**

~~~
Downloading ruby-3.0.0.tar.gz...
-> https://cache.ruby-lang.org/pub/ruby/3.0/ruby-3.0.0.tar.gz
Installing ruby-3.0.0...

BUILD FAILED (Ubuntu 20.04 using ruby-build 20201225)

Inspect or clean up the working tree at /tmp/ruby-build.20210110021750.214336.MBU7kX
Results logged to /tmp/ruby-build.20210110021750.214336.log

Last 10 log lines:
installing bundled gem cache:       /home/user/ruby-build/bin/./ruby-test/lib/ruby/gems/3.0.0/cache
The Ruby openssl extension was not compiled.
ERROR: Ruby install aborted due to missing extensions
Try running `apt-get install -y libssl-dev` to fetch missing dependencies.

Configure options used:
  --prefix=/home/user/ruby-build/bin/./ruby-test
  --enable-shared
  LDFLAGS=-L/home/user/ruby-build/bin/./ruby-test/lib 
  CPPFLAGS=-I/home/user/ruby-build/bin/./ruby-test/include
~~~

There were a lot of great suggestions in the [OP](https://github.com/rbenv/ruby-build/issues/1215) on addressing the issue back then.  Tried all of them, but no luck:

* https://github.com/rbenv/ruby-build/issues/1215#issuecomment-399687588

Debian-specific, no package with that name in Ubuntu

* https://github.com/rbenv/ruby-build/issues/1215#issuecomment-462122470

Was already installed.  Did `apt install --reinstall` for both `libreadline-dev` and `libssl-dev` as well

* https://github.com/rbenv/ruby-build/issues/1215#issuecomment-571452147

Tried this with `openssl-1.1.1i.tar.gz` ([latest](https://www.openssl.org/source/))

**Comment by NathanZook** [comment]
I just updated to Mint 19, which I believe is based on 20.04.  According to Synaptic, libssl-dev is present (and what you want for modern rubies).  Note that there is an incompatibility between libssl-dev and libssl1.0-dev.  I installed the latter to build some older rubies, and had trouble removing it.  I'm not sure why it worked when I marked it for uninstall the second time.

**Comment by eregon** [comment]
We need the full log here to see the compilation error.

**Comment by affix** [comment]
I get this exact same error on Fedora 34 with any Version of Ruby

```
» rbenv install 3.1.0-dev  
Cloning https://github.com/ruby/ruby.git...
Installing ruby-master...
ruby-build: using readline from homebrew

BUILD FAILED (Fedora 34 using ruby-build 20210707)

Inspect or clean up the working tree at /tmp/ruby-build.20210709174027.1619271.9XOBar
Results logged to /tmp/ruby-build.20210709174027.1619271.log

Last 10 log lines:
installing bundled gem cache:       /home/affix/.rbenv/versions/3.1.0-dev/lib/ruby/gems/3.1.0/cache
The Ruby openssl extension was not compiled.
ERROR: Ruby install aborted due to missing extensions
Try running `yum install -y openssl-devel` to fetch missing dependencies.

Configure options used:
  --prefix=/home/affix/.rbenv/versions/3.1.0-dev
  --with-readline-dir=/home/linuxbrew/.linuxbrew/opt/readline
  LDFLAGS=-L/home/affix/.rbenv/versions/3.1.0-dev/lib 
  CPPFLAGS=-I/home/affix/.rbenv/versions/3.1.0-dev/include 

```

Full Log : https://pastebin.com/CtA5cUHR

**Comment by dre-hh** [comment]
It fails because the code was intended for use with homebrew and not with linuxbrew.
There is a similar issue in python-build (pyenv)
https://github.com/pyenv/pyenv/issues/1479#issuecomment-610683526
It seems like both libs would need a redesign to ignure  homebrew on linux 
The workaround for now is temporarily removing homebrew form the path

**Comment by rogeraraujo90** [comment]
is this still a problem without a solution for Ubuntu 22.04 and Ruby 3.2? I'm getting this error and, removing homebrew from the path has worked, but since this is a very old discussion I'm wondering if we might have a definitive solution.

**Reply by mislav** [reply]
@rogeraraujo90 Note that the OP never posted the full build log, so the original cause of the failure was not fully known. I think it's safe to say that whichever failure you are getting today, it's most likely not related to the one experienced in January 2021, since Ubuntu, ruby-build, and Ruby itself have all shipped a ton of updates since then.

If you have a problem that you can reasonably replicate on an Ubuntu machine with the latest ruby-build version, then please open a new discussion explaining the problem, and _please_ include build log as a Gist because without one we can't really know what caused the error.
